dstep/translator/Preprocessor.d

```diff
@@ -0,0 +1,935 @@
+/**
+ * Copyright: Copyright (c) 2016 Wojciech Szęszoł. All rights reserved.
+ * Authors: Wojciech Szęszoł
+ * Version: Initial created: Jul 08, 2016
+ * License: $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0)
+ */
+module dstep.translator.Preprocessor;
+
+import std.array;
+
+import clang.Cursor;
+import clang.Index;
+import clang.SourceRange;
+import clang.Token;
+import clang.TranslationUnit;
+
+public import dstep.translator.MacroDefinition;
+
+class ConditionalDirective : Directive
+{
+    Expression condition;
+    Directive[] branches;
+    Directive endif;
+}
+
+class PragmaDirective : Directive
+{
+}
+
+class UndefDirective : Directive
+{
+    string identifier;
+}
+
+struct TokenizedDirectiveRange
+{
+    string source;
+    Token[] tokens;
+    Token[] result;
+
+    private void findNext()
+    {
+        if (1 < tokens.length &&
+            tokens[0].spelling == "#" &&
+            isDirective(tokens[1]))
+            return;
+
+        while (2 < tokens.length &&
+            (tokens[1].spelling != "#" ||
+            tokens[0].location.line >= tokens[1].location.line ||
+            !isDirective(tokens[2])))
+            tokens.popFront();
+
+        if (2 < tokens.length)
+            tokens.popFront();
+        else
+            tokens = Token[].init;
+    }
+
+    private size_t countEscaped(string range)
+    {
+        return 0;
+    }
+
+    private void yield()
+    {
+        result = Token[].init;
+
+        size_t itr = 0;
+
+        while (itr + 1 < tokens.length)
+        {
+            auto loc0 = tokens[itr].location;
+            auto loc1 = tokens[itr + 1].location;
+
+            ptrdiff_t diff = loc1.line - loc0.line;
+
+            if (0 < diff &&
+                (source == null ||
+                countEscaped(source[loc0.offset .. loc1.offset]) < diff))
+                break;
+
+            ++itr;
+        }
+
+        if (itr + 1 < tokens.length)
+        {
+            result = tokens[0 .. itr + 1];
+            tokens = tokens[itr + 1 .. $];
+        }
+        else
+        {
+            result = tokens;
+            tokens = Token[].init;
+        }
+    }
+
+    private static bool isDirective(Token token)
+    {
+        switch (token.spelling)
+        {
+            case "elif":
+            case "else":
+            case "endif":
+            case "error":
+            case "define":
+            case "if":
+            case "ifdef":
+            case "ifndef":
+            case "include":
+            case "line":
+            case "pragma":
+            case "undef":
+                return true;
+
+            default:
+                return false;
+        }
+    }
+
+    this(Token[] tokens, string source = null)
```

Comment:
Prefer to put the constructor first of the functions and private functions last.